### PR TITLE
SAK-41402: Site Info > Duplicate Site > provide meaningful email alias

### PR DESCRIPTION
**Pull request by bjones86**
https://jira.sakaiproject.org/browse/SAK-41402

See screenshot on the JIRA. Wording of introduced error message up for discussion.

Currently when you duplicate a site which contains the Email Archive tool, the resulting duplicated site's Email Archive tool will have the site's email alias set to the UUID of the site. This isn't very intuitive for the user; they may not immediately notice the alias, and when they do they may not understand what the UUID is or why it was set this way.

The linked PR proposes changing this behaviour to provide more control for the user to determine what the email alias should be. If the originating site contains the Email Archive tool, the user will be presented with a new, dedicated text box to enter their desired email alias.

If the user does not provide any input into the email alias field, they are presented with an appropriate error message which blocks the duplication process. If the user provides an email alias which is already in use by another site, the user is presented with an appropriate error message which blocks the duplication process. Once the user has provided a valid, unique email alias, the duplication process can be completed normally.

During testing of this PR, I also noticed that the error messaging was not properly implemented. If any error besides the first one in the code (which checks if a site title was provided) is triggered, the site duplication process proceeds as if no errors were encountered. This PR fixes this bug so that proper error messaging is displayed to the user before the duplication process continues.

I also fixed a bug where the user supplied site title would not be pre-populated in the form in the event of an error message being displayed (loss of user data).

### CODE REVIEW

**Comment by bjones86** [comment]
The diff on this looks more complicated than it actually is. Most of it is just moving blocks of code around, and indenting them due to proper checking of error message present in the `state`.

File: site-manage/site-manage-tool/tool/src/java/org/sakaiproject/site/tool/SiteAction.java [hunk 1]
```diff
@@ -9745,215 +9746,258 @@ private void actionForTemplate(String direction, int index,
...
-									List<SitePage> rmPageList = new ArrayList<>();
+								Site site = SiteService.addSite(newSiteId,
+										getStateSite(state));
 
-									// Check if each tool is stealthed; if so, queue for removal
-									for (SitePage page : pageList) {
-										List<ToolConfiguration> pageToolList = page.getTools();
-										if (CollectionUtils.isNotEmpty(pageToolList)) {
-											List<ToolConfiguration> rmToolList = new ArrayList<>();
+								String newEmailID = titleStripped;
```
Comment (bjones86) [inline site-manage/site-manage-tool/tool/src/java/org/sakaiproject/site/tool/SiteAction.java]: New functionality change starts here.

File: site-manage/site-manage-tool/tool/src/java/org/sakaiproject/site/tool/SiteAction.java [hunk 2]
```diff
@@ -9745,215 +9746,258 @@ private void actionForTemplate(String direction, int index,
...
+														aliasService.setAlias(newEmailID, channelReference);
+													} catch (IdUsedException ex) {
+													} catch (IdInvalidException ex) {
+													} catch (PermissionException ex) {}
 												}
-											}
+											} catch (PermissionException ex) {}
+										}
+									}
+								}
```
Comment (bjones86) [inline site-manage/site-manage-tool/tool/src/java/org/sakaiproject/site/tool/SiteAction.java]: New functionality change ends here.

File: site-manage/site-manage-tool/tool/src/java/org/sakaiproject/site/tool/SiteAction.java [hunk 3]
```diff
@@ -9749,215 +9762,262 @@ private void actionForTemplate(String direction, int index,
...
+								// SAK-20797
+								long oldSiteQuota = this.getSiteSpecificQuota(oldSiteId);
+
+								// Create the duplicate site
+								Site site = SiteService.addSite(newSiteId, getStateSite(state));
+								site.setTitle(titleStripped);
+
+								// If the site contains the Email Archive tool, we're going to check for valid/unique email address;
+								// this requires the duplicate site to already exist...
+								if (state.getAttribute(STATE_DUP_SITE_HAS_EMAIL_ARCHIVE) != null) {
```
Comment (bjones86) [inline site-manage/site-manage-tool/tool/src/java/org/sakaiproject/site/tool/SiteAction.java]: New code starts here...

File: site-manage/site-manage-tool/tool/src/java/org/sakaiproject/site/tool/SiteAction.java [hunk 4]
```diff
@@ -9749,215 +9762,262 @@ private void actionForTemplate(String direction, int index,
...
+													// If aliasService.getTarget() throws this, it's all good: email alias not yet in use, so we set it
+													try {
+														aliasService.setAlias(newEmailID, channelReference);
+													} catch (Exception e) {
+														addAlert(state, rb.getFormattedMessage("unexpectedError", new Object[] {ServerConfigurationService.getString("mail.support")}));
+														deleteTempDupSiteOnError(site);
+													}
+												} catch (Exception ex) {
+													addAlert(state, rb.getFormattedMessage("unexpectedError", new Object[] {ServerConfigurationService.getString("mail.support")}));
+													deleteTempDupSiteOnError(site);
```
Comment (bjones86) [inline site-manage/site-manage-tool/tool/src/java/org/sakaiproject/site/tool/SiteAction.java]: ends here, with the exception of the new `deleteTempDupSiteOnError()` defined below

File: site-manage/site-manage-tool/tool/src/java/org/sakaiproject/site/tool/SiteAction.java [hunk 5]
```diff
@@ -9749,215 +9762,262 @@ private void actionForTemplate(String direction, int index,
...
-							// notification of the start/end of a site duplication.
-							String sourceSiteRef = null;
-							try {
-								Site sourceSite = SiteService.getSite(oldSiteId);
-								sourceSiteRef = sourceSite.getReference();
-								
-							} catch (IdUnusedException e) {
-								log.warn(this + ".actionForTemplate; case29: invalid source siteId: "+oldSiteId);
-								return;
+							String newSiteId = null;
```
Comment (ottenhoff) [inline site-manage/site-manage-tool/tool/src/java/org/sakaiproject/site/tool/SiteAction.java]: easy opportunity here to get of the else part

File: site-manage/site-manage-tool/tool/src/java/org/sakaiproject/site/tool/SiteAction.java [hunk 6]
```diff
@@ -9749,215 +9762,262 @@ private void actionForTemplate(String direction, int index,
...
+
+								// Create the duplicate site
+								Site site = SiteService.addSite(newSiteId, getStateSite(state));
+								site.setTitle(titleStripped);
+
+								// If the site contains the Email Archive tool, we're going to check for valid/unique email address;
+								// this requires the duplicate site to already exist...
+								if (state.getAttribute(STATE_DUP_SITE_HAS_EMAIL_ARCHIVE) != null) {
+									String newEmailID = StringUtils.trimToNull(params.getString("emailAddress"));
+									if (StringUtils.isBlank(newEmailID)) {
```
Comment (ottenhoff) [inline site-manage/site-manage-tool/tool/src/java/org/sakaiproject/site/tool/SiteAction.java]: already trimmed to null directly above

File: site-manage/site-manage-tool/tool/src/java/org/sakaiproject/site/tool/SiteAction.java [hunk 7]
```diff
@@ -9749,215 +9762,262 @@ private void actionForTemplate(String direction, int index,
...
+										addAlert(state, rb.getString("java.emailarchive"));
+										deleteTempDupSiteOnError(site);
+									} else {
+										state.setAttribute(STATE_TOOL_EMAIL_ADDRESS, newEmailID);
+										if (StringUtils.isNotBlank(newEmailID)) {
```
Comment (ottenhoff) [inline site-manage/site-manage-tool/tool/src/java/org/sakaiproject/site/tool/SiteAction.java]: no way you can be blank here, right?

File: site-manage/site-manage-tool/tool/src/java/org/sakaiproject/site/tool/SiteAction.java (hunk 5 above)
Comment (bjones86) [reply site-manage/site-manage-tool/tool/src/java/org/sakaiproject/site/tool/SiteAction.java]: This code wasn't written as part of this patch, just moved from one location to another. If need be, I can refactor the existing code...

File: site-manage/site-manage-tool/tool/src/java/org/sakaiproject/site/tool/SiteAction.java [hunk 8]
```diff
@@ -10173,6 +10233,38 @@ private void actionForTemplate(String direction, int index,
...
+			SecurityService.pushAdvisor(yesMan);
+
+			// Hard delete. Call upon all implementing services to hard delete their own content
+			doHardDelete(tempDupSite.getId());
+			// The service never deletes the site unless its already softly deleted
+			tempDupSite.setSoftlyDeleted(true);
+
+			// Now hard delete the site
+			SiteService.removeSite(tempDupSite);
+		} catch (Exception e) {
```
Comment (ottenhoff) [inline site-manage/site-manage-tool/tool/src/java/org/sakaiproject/site/tool/SiteAction.java]: should at least log one line here

File: site-manage/site-manage-tool/tool/src/java/org/sakaiproject/site/tool/SiteAction.java (hunk 6 above)
Comment (bjones86) [reply site-manage/site-manage-tool/tool/src/java/org/sakaiproject/site/tool/SiteAction.java]: ditto here, this is existing code that was just moved around.

File: site-manage/site-manage-tool/tool/src/java/org/sakaiproject/site/tool/SiteAction.java (hunk 8 above)
Comment (bjones86) [reply site-manage/site-manage-tool/tool/src/java/org/sakaiproject/site/tool/SiteAction.java]: ditto here, existing code that was just moved.

**Pull request merged**